modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -1214,6 +1232,51 @@ else if (spiState == LOOPBACK_PROBLEM) {
             .end();
     }
 
+    /**
+     * Schedules a maintenance task to clear cache folders based on the exception message.
+     *
+     * @param checkFailedMsg Exception message.
+     */
+    private void scheduleMaintenanceTaskToClearCacheFolders(TcpDiscoveryCheckFailedMessage checkFailedMsg) {
+        try {
+            ArrayList<CacheConfiguration> cacheCfgs = new ArrayList<>();
+            GridCacheSharedContext sharedCacheContext = gridKernalContext().cache().context();
+
+            for (String cacheName : checkFailedMsg.error()
+                .replaceFirst("^.*\\[", "")
+                .replaceFirst("\\].*", "")
+                .split(", ")) {
+                CacheConfiguration cc = sharedCacheContext.cacheContext(CU.cacheId(cacheName)).config();
+
+                cacheCfgs.add(cc);
+            }
+
+            FilePageStoreManager pageStore = (FilePageStoreManager)sharedCacheContext.pageStore();
+
+            String params = cacheCfgs.stream()
+                .map(ccfg -> pageStore.cacheWorkDir(ccfg).getName())
+                .collect(Collectors.joining(File.separator));
+
+            log.warning("The node won't join the cluster, because it has some several caches, that were removed" +
```

Comment:
There are some grammar and word-usage mistakes in this message. For instance, "data" isn't "connected with the caches", it is better to say "left from the caches" or "related to the caches".

"The node is going to enter maintenance mode after restart and will clear files of stale caches automatically".

"When the task is finished, node has to be restarted once again to leave maintenance mode and join the cluster back".

Something like that.